### ISSUE TITLE
Roundtrip encoding/decoding of unknown fields

### ISSUE THREAD

**Issue by bobergj** [tomas-abrahamsson/gpb#189]
> Originally, proto3 messages always discarded unknown fields during parsing, but in version 3.5 we reintroduced the preservation of unknown fields to match the proto2 behavior. In versions 3.5 and later, unknown fields are retained during parsing and included in the serialized output.

https://developers.google.com/protocol-buffers/docs/proto3#unknowns

The use-case here is being able to evolve the peers communicating through an intermediary - that proxies or partially processes the messages - without having to update the intermediary.
The intermediary can even mutate the message fields it recognizes, without affecting the unknown fields.

My impression is that in gpb currently does not support this? I am using the  record output option for code generation, perhaps with maps the story is different?

In the Swift protocol-buffer library I am also using, unknown fields get represented as a binary blob in the decoded message. Every generated message struct has a property:
```
    public var unknownFields = SwiftProtobuf.UnknownStorage()
```
where UnknownStorage is simply a byte sequence that gets appended as unknown fields are encountered.
I am thinking that in gpb this could be similarly implemented by adding a _unknown_fields field to each message record.
What do you think?

**Comment by tomas-abrahamsson** [comment]
> My impression is that in gpb currently does not support this? I

This is true in gpb for both records and maps. I think it could be done, and can look into it.

**Comment by tomas-abrahamsson** [comment]
I have pushed to a branch `preserve-unknown-fields` to support this. I have not yet looked into support for json and nif for this (don't know if you would need that?). But for records and maps it should work, if you want to try it out. The branch may still change, and once merged to master, I will eventually remove the branch.

**Comment by bobergj** [comment]
Thank you! I haven't had a chance to test it yet, but the branch looks promising.
I see there is an option to opt-in to preserving unknown fields. In the Google reference implementations for the various languages, preserving is default, and there is an option for discarding unknown fields. Perhaps gpb should do the same?

 > I have not yet looked into support for json and nif for this (don't know if you would need that?)

No, I don't use those options.
Also, the language guide says a proto3 json parser should reject unknown fields by default - as in, fail to parse the whole message - but may have an option to skip them. Preservation of unknowns is not mentioned for the json encoding.

**Comment by tomas-abrahamsson** [comment]
> In the Google reference implementations ... preserving is default ...

Yes, this is a valid concern. I noticed and was considering this, but I am leaning towards keeping it opt-in, because it would otherwise add fields to records and maps, and that could potentially cause breakage. An example could be a messasge pretty-printer that traverses records and message definitions to format a string with field names, but suddenly finds a field with a new type `unknown` in each message.

Thanks for the info on handling in json. I figure the problem is that when parsing an unknown json number or integer field value, it is generally not possible to know whether it is a 32- or 64-bit or varint value, nor its field number (the `=⟨n⟩`) but when encoding using the protobuf wire-format (also unknowns) these are required pieces of information.

**Comment by tomas-abrahamsson** [comment]
I've added nif support (pushed to the `preserve-unknown-fields`) Just wanted to let you know that I'm generally happy now with the shape of this branch and that I intend to merge and ship it soon, like within a day or so hopefully.

**Comment by bobergj** [comment]
Sounds good! I have already deployed your branch (although the revision before the nif support) in a staging environment, and I can confirm it works as it should. I have clients, using the swift-protobuf library, successfully exchanging messages containing properties unknown to the server - which uses gpb.
Thanks for your work on this.

**Comment by tomas-abrahamsson** [comment]
Thanks for the update! I've now merged and released 4.16.0, so I will close this. Should you discover anything, feel free to reopen or open a new issue!
